### PR TITLE
Frontend: friends page layout + friends list display

### PR DESCRIPTION
**Pull request by darrenvong**
Lays out the page for #291, with the friend list display completed. Since there's currently no designs on this, I'm going along with the alpha app to make some progress, hence the new "friends" link added in the navigation bar.

Doing listing of friend requests and listing of friend requests you sent separately so the PR doesn't become a monster. The components for "friend requests" and "friend requests you sent" are there for now as a placeholder on the page.

Despite the strong reaction of checking for component mounted state being a "hack", I don't see it being that bad once abstracted away into their own hooks as in this PR. Also kinda doing this on purpose to avoid writing too much Redux code unless it's absolutely needed while we're still deciding if we are going to stick with Redux.

### CODE REVIEW

**Comment by amdwit** [comment]
I think this should be used for (part of) the design:
https://couchers.invisionapp.com/console/share/2W1JUDHBZ7/483411816

**Comment by lucaslcode** [comment]
> Despite the strong reaction of checking for component mounted state being a "hack", I don't see it being that bad

I agree with this

**Comment by darrenvong** [comment]
> I think this should be used for (part of) the design:
> https://couchers.invisionapp.com/console/share/2W1JUDHBZ7/483411816

Oh I missed this! Although this seems to split connections by list types, which I suspect we don't have the APIs for creating these connection list type yet.

Not sure what the best way to lay this out under this connection page. Two ideas I got on top of my head:
- Put each of the panels I got (copied from the alpha app) as three tabs: Friends, Pending, Requests Sent
- Combine all three of those under a single "Friends" tab (since they are the same type of connection) and show them in the same layout as what I got in the PR.

Thoughts?

**Comment by darrenvong** [comment]
> > Despite the strong reaction of checking for component mounted state being a "hack", I don't see it being that bad
> 
> I agree with this

I think I have a similar "bias" as you towards polluting global states haha. It feels wrong to dump a loading state there when only a single component cares about it and the only reason we're doing it is to solve setting state on a potentially unmounted component. This "hack"/check feels like a ok compromise to keep things simple for me imo. I think something similar is even suggested by Dan from the React core team in [his blog post](https://overreacted.io/a-complete-guide-to-useeffect/#speaking-of-race-conditions), so it can't be that bad... right? :sweat_smile:

**Comment by amdwit** [comment]
Your decision to depart from the 'actions up data down' paradigm causes the following problems: friendIds is not easily persistable, the isMounted hack is needed causing a lot of extra boiler plate in the component. That paradigm is there for a reason and who knows what more problems will show up. Couldn't this PR be a nice demo for react-query? I think when the hydration functionality of the queryCache is used then the persisting problem is solved.

File: app/frontend/src/features/friends/FriendRequestsSent.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  },
+  errorAlert: {
+    borderRadius: 0,
+  },
+  header: {
+    marginLeft: theme.spacing(1),
+    marginBottom: theme.spacing(2),
+    fontWeight: theme.typography.fontWeightBold,
+  },
+}));
```
Comment (lucaslcode) [inline app/frontend/src/features/friends/FriendRequestsSent.tsx]: unused?

File: app/frontend/src/features/friends/FriendsPage.tsx [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      <PageTitle>Friends</PageTitle>
+      <Grid container spacing={2}>
+        <Grid item xs={12} md={6}>
+          <FriendList />
+        </Grid>
+        <Grid item xs={12} md={6}>
+          <FriendRequests />
+        </Grid>
+        <Grid item xs={12} md={6}>
+          <FriendRequestsSent />
```
Comment (lucaslcode) [inline app/frontend/src/features/friends/FriendsPage.tsx]: Are these supposed to be md=4?

File: app/frontend/src/features/messages/surfing/HostRequestView.tsx [hunk 3]
```diff
@@ -37,8 +37,10 @@ export default function HostRequestView({ hostRequest }: HostRequestViewProps) {
   }, [hostRequest.hostRequestId]);
 
   useEffect(() => {
-    dispatch(fetchUsers({ userIds: [hostRequest.fromUserId, hostRequest.toUserId] }));
-  }, [hostRequest.fromUserId, hostRequest.toUserId]);
+    dispatch(
+      fetchUsers({ userIds: [hostRequest.fromUserId, hostRequest.toUserId] })
+    );
+  }, [dispatch, hostRequest.fromUserId, hostRequest.toUserId]);
```
Comment (lucaslcode) [inline app/frontend/src/features/messages/surfing/HostRequestView.tsx:43]: I'm not too fussed but some other devs prefer unrelated things to be in a different PR. (Or maybe a seperate commit)

File: app/frontend/src/utils/hooks.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    return () => {
+      isMounted.current = false;
+    };
+  }, []);
+
+  return isMounted;
+}
+
+function useSafeState<State>(
+  isMounted: MutableRefObject<boolean>,
```
Comment (lucaslcode) [inline app/frontend/src/utils/hooks.ts:25]: Why not put `useIsMounted` in this function? Then it doesn't need this argument.

**Review by lucaslcode — CHANGES_REQUESTED** [review]

**Comment by lucaslcode** [comment]
> Your decision to depart from the 'actions up data down' paradigm causes the following problems

I don't see this as being violated here. There is no state relevant to higher components.

> friendIds is not easily persistable

This is true, but we don't need to persist it. If we did, then it would make sense to put in the redux store.

> isMounted hack is needed causing a lot of extra boiler plate in the component

With the custom hooks, there isn't much extra boilerplate, and less than putting it in the redux store.

Another thing I thought of, perhaps there's some conflict between separation of views and model layer (not sure if that's the right terminology) and keeping code related to individual components together.

File: app/frontend/src/utils/hooks.ts (hunk 4 above)
Comment (amdwit) [reply app/frontend/src/utils/hooks.ts:25]: I think this is not done because of repetition.

**Comment by amdwit** [comment]
Maybe we must also decide how much we want to persist. I thought that users might need to access previously accessed data even when offline. That would mean we need to persist friendIds (and users). What do you think about offline usage?

**Comment by amdwit** [comment]
> I don't see this as being violated here.

Then we have different ideas about. For me it is mainly about components triggering actions at a data owner. The components react to the data owner. I think a central data owner is best. If we do not want to work that way then I prefer to know. Which I do now.

**Comment by amdwit** [comment]
> and less than putting it in the redux store.

But more than putting it in mobx/mst. (no set functions needed, and no isMounted) . But using a custom hook like useQuery which calls the fetch logic and returns loading and error and has the useSafeState logic, would be much cleaner in this case.

**Comment by amdwit** [comment]
> Another thing I thought of, perhaps there's some conflict between separation of views and model layer (not sure if that's the right terminology) and keeping code related to individual components together.

Indeed, and the big question there is about persisting. If this is needed and we won't let the service worker cache api calls  then we should go for option 1.

File: app/frontend/src/utils/hooks.ts (hunk 4 above)
Comment (darrenvong) [reply app/frontend/src/utils/hooks.ts:25]: Yeah it's mainly to stop the need for the `isMounted` code to run more than once within a component's lifecycle.

File: app/frontend/src/features/friends/FriendRequestsSent.tsx (hunk 1 above)
Comment (darrenvong) [reply app/frontend/src/features/friends/FriendRequestsSent.tsx]: Ahh yeah forgot to delete this when I moved this to be shared styles, good spot :+1:

File: app/frontend/src/features/friends/FriendsPage.tsx (hunk 2 above)
Comment (darrenvong) [reply app/frontend/src/features/friends/FriendsPage.tsx]: I did `md=6` here as the [WIP hi-fi desktop wireframes here](https://www.figma.com/file/wYr6z7KQpoIIN0A97jD7kf/Couchers-Platform-Design?node-id=680%3A1) seem to suggest we want to show two tiles per row. So maybe we should stick with it? :shrug:

File: app/frontend/src/features/messages/surfing/HostRequestView.tsx (hunk 3 above)
Comment (darrenvong) [reply app/frontend/src/features/messages/surfing/HostRequestView.tsx:43]: Good point, it does make PRs more difficult to review otherwise if this is larger. I guess since this is a small addition that doesn't change much than pleasing the linter... I'm going to leave it in for this time but I'll try not do that in future!

**Comment by darrenvong** [comment]
> Another thing I thought of, perhaps there's some conflict between separation of views and model layer (not sure if that's the right terminology) and keeping code related to individual components together.

This is a good point. For states that don't need to be persisted, I'd personally prefer to keep related code together in the component - a bunch of `setState`s feels easier to follow for me and it's nice to be able to see all related logic without having to open multiple files just to get a full picture of how everything works.

> > and less than putting it in the redux store.
> 
> But more than putting it in mobx/mst. (no set functions needed, and no isMounted) . But using a custom hook like useQuery which calls the fetch logic and returns loading and error and has the useSafeState logic, would be much cleaner in this case.

But then we have to introduce state external to what's available out of the box from React, even if we don't need to persist. If custom hooks solve the problem, I think there's merit in keeping things in React since that's less things to learn for new people in the project. Also it's only two custom hooks which is much lighter weight than bringing in another state management library just for this scenario.

**Comment by darrenvong** [comment]
Closing since this will be covered by the demo PR using React Query instead as agreed in the last meeting.
